[FEATURE] improve performance by eldariont

Irallia opened the pull request:
This PR replaces https://github.com/seqan/iGenVar/pull/142 

> This PR contains several changes to
> 
>     * the variant detection from split alignments,
> 
>     * the hierarchical clustering and
> 
>     * the variant output logic
> 
> 
> with the goal of improving the performance of our caller in the HG002 SV benchmark.
> 
> Currently, this is still a draft and we should measure the change in benchmark results it causes (after merging #141). I also need to fix the tests.

# Performance improvement:
before
[CallerComparisonPlot.pdf](https://github.com/seqan/iGenVar/files/7184928/CallerComparisonPlot.pdf)

after
[CallerComparisonPlot.pdf](https://github.com/seqan/iGenVar/files/7184943/CallerComparisonPlot.pdf)

# Parameter changes:
hierarchical_clustering_cutoff:
![hierarchical_clustering_cutoff results all](https://user-images.githubusercontent.com/5746285/133780821-0350454a-223d-488e-a697-5ec208597a8a.png)
max_tol_deleted_length:
![max_tol_deleted_length results all](https://user-images.githubusercontent.com/5746285/133780822-e7ed9f15-2461-4b0c-9308-2788d421e7af.png)
partition_max_distance:
![partition_max_distance results all](https://user-images.githubusercontent.com/5746285/133780825-ee40f0b7-86b1-4ad4-8d7e-f4058fa45797.png)

SGSSGene commented on src/modules/clustering/hierarchical_clustering_method.cpp:
```diff
@@ -87,25 +91,54 @@ std::vector<std::vector<Junction>> split_partition_based_on_mate2(std::vector<Ju
+                (lhs_directed_size > 0 && rhs_directed_size > 0))
+            {
+                double position_distance = std::abs(lhs.get_mate1().position - rhs.get_mate1().position) / 1000.0;
+                // TODO (irallia 01.09.2021): std::abs((int)(lhs.get_tandem_dup_count() - rhs.get_tandem_dup_count()))
```
What about this todo?

Irallia replied:
Thats a todo for my next tandem duplication PR :)

SGSSGene commented on src/variant_detection/variant_output.cpp:
```diff
@@ -32,10 +32,11 @@ void find_and_output_variants(std::map<std::string, int32_t> & references_length
                         //Deletion
-                        if (distance >= args.min_var_length &&
-                            distance <= args.max_var_length &&
+                        if (sv_length <= -args.min_var_length &&
+                            sv_length >= -args.max_var_length &&
```
These lines are very confusing... is sv_length a negative number?
I would suggest to write:
```suggestion
                        if (-sv_length >= args.min_var_length &&
                            -sv_length <= args.max_var_length &&
```

SGSSGene commented on src/variant_detection/variant_output.cpp:
```diff
@@ -45,15 +46,16 @@ void find_and_output_variants(std::map<std::string, int32_t> & references_length
                             tmp.set_pos(mate1_pos + 1);
-                            tmp.add_info("SVLEN", std::to_string(-distance + 1));
+                            tmp.add_info("SVLEN", std::to_string(sv_length));
                             // Increment end by 1 because VCF is 1-based
                             // Decrement end by 1 because deletion ends one base before mate2 begins
```
These two comments seem to be outdated. There is no increment/decrement

Irallia replied:
Yes, but only because they cancel each other out.

SGSSGene replied:
:facepalm: yes, they cancel each other out, perfectly good comment :-)

SGSSGene reviewed (changes requested):
Looks good, just some small stuff.

SGSSGene reviewed (approved):
nice :-)

Irallia commented:
@joergi-w : I would ignore the codecoverage loss. We can also go through parts of this PR on Wednesday if some things are not clear to you. It's a lot of stuff...
It is the code that was left from @eldariont . And since it actually brings some improvement, I rebased and added it.

joergi-w commented on include/modules/clustering/hierarchical_clustering_method.hpp:
```diff
@@ -32,15 +37,18 @@ std::vector<std::vector<Junction>> split_partition_based_on_mate2(std::vector<Ju
  * \param[in] rhs - right side junction
  */
-int junction_distance(Junction const & lhs, Junction const & rhs);
+double junction_distance(Junction const & lhs, Junction const & rhs);
```
missing `\returns` in doxygen documentation (+ 3x in this file)

joergi-w commented on include/iGenVar.hpp:
```diff
@@ -27,11 +27,13 @@ struct cmd_arguments
     /* -q */ int32_t min_qual = 1;
 // Clustering specifications:
-    /* -w */ double hierarchical_clustering_cutoff = 100;
+    /* -p */ int32_t partition_max_distance = 1000;
```
If the user must pass non-negative values, just use an unsigned type. This automatically puts `(unsigned 32 bit integer)` to the help screen and prints a helpful error message if a user passes a negative value nonetheless.
```suggestion
    /* -p */ uint32_t partition_max_distance = 1000;
```

Irallia replied:
I opened an issue for that: https://github.com/seqan/iGenVar/issues/161

joergi-w commented on src/iGenVar.cpp:
```diff
@@ -94,6 +94,10 @@ void initialize_argument_parser(seqan3::argument_parser & parser, cmd_arguments
                       seqan3::option_spec::advanced);
+    parser.add_option(args.max_tol_deleted_length, 'e', "max_tol_deleted_length",
+                      "Specify what should be the longest tolerated deleted sequence at sites of non-DEL SVs. "
+                      "This value needs to be non-negative.",
```
this extra comment wouldn't be necessary with unsigned parameter variable

Irallia replied:
-> https://github.com/seqan/iGenVar/issues/161

joergi-w commented on src/modules/clustering/hierarchical_clustering_method.cpp:
```diff
@@ -25,13 +26,14 @@ std::vector<std::vector<Junction>> partition_junctions(std::vector<Junction> con
             if (junction.get_mate1().seq_name != current_partition.back().get_mate1().seq_name ||
                 junction.get_mate1().orientation != current_partition.back().get_mate1().orientation ||
-                abs(junction.get_mate1().position - current_partition.back().get_mate1().position) > 50)
+                abs(junction.get_mate1().position - current_partition.back().get_mate1().position)
```
doesn't this need `std::abs`? (I'm not sure, but if it works on every machine - fine!)
Update: at least below you use `std::`, so decide on one version please

joergi-w commented on src/variant_detection/variant_detection.cpp:
```diff
@@ -90,11 +90,13 @@ void detect_junctions_in_short_reads_sam_file(std::vector<Junction> & junctions,
+        if (gVerbose)
         {
-            seqan3::debug_stream << num_good << " good alignments from short read file." << std::endl;
+            num_good++;
```
avoid a copy:
```suggestion
            ++num_good;
```

joergi-w commented on src/variant_detection/variant_detection.cpp:
```diff
@@ -174,10 +176,13 @@ void detect_junctions_in_long_reads_sam_file(std::vector<Junction> & junctions,
+        if (gVerbose)
         {
-            seqan3::debug_stream << num_good << " good alignments from long read file." << std::endl;
+            num_good++;
```
same here
